util/webhook/certs/certmanager.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package certs
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"strconv"
+	"time"
+
+	adminregv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"knative.dev/pkg/webhook/certificates/resources"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// New returns a new instance of a CertificateManager.
+func New(options Options) (*CertificateManager, error) {
+	if err := options.defaultOpts(); err != nil {
+		return nil, err
+	}
+	return &CertificateManager{opts: options}, nil
+}
+
+// CertificateManager creates and rotates certificates required by a controller manager's webhook server.
+type CertificateManager struct {
+	opts Options
+}
+
+// Start starts certificate management for a controller manager's webhooks.
+// This method calls os.Exit when an error is encountered.
+func (cm *CertificateManager) Start(ctx context.Context) error {
+	go func() {
+		if err := cm.start(ctx); err != nil {
+			cm.opts.Logger.Error(err, "unexpected failure in starting certificate manager")
+			os.Exit(1)
```

Comment:
Is it okay to stop the controller even for transient network failures such as transient failure to reach the API server?

Reply:
Yes, it's okay because the deployment will make sure pod is restarted when it exits.